plugins/inputs/internet_speed/internet_speed.go

```diff
@@ -33,63 +39,79 @@ func (*InternetSpeed) SampleConfig() string {
 func (is *InternetSpeed) Init() error {
 	is.MemorySavingMode = is.MemorySavingMode || is.EnableFileDownload
 
+	var err error
+	is.serverFilter, err = filter.NewIncludeExcludeFilter(is.ServerIDInclude, is.ServerIDExclude)
+	if err != nil {
+		return fmt.Errorf("error compiling server ID filters: %v", err)
+	}
+
 	return nil
 }
 
 func (is *InternetSpeed) Gather(acc telegraf.Accumulator) error {
-	// Get closest server
-	s := is.serverCache
-	if s == nil {
-		user, err := speedtest.FetchUserInfo()
-		if err != nil {
-			return fmt.Errorf("fetching user info failed: %v", err)
-		}
-		serverList, err := speedtest.FetchServers(user)
-		if err != nil {
-			return fmt.Errorf("fetching server list failed: %v", err)
-		}
-		if len(serverList) < 1 {
-			return fmt.Errorf("no servers found")
-		}
-		s = serverList[0]
-		is.Log.Debugf("Found server: %v", s)
-		if is.Cache {
-			is.serverCache = s
+	// if not caching, go find closest server each time
+	if !is.Cache || is.server == nil {
+		if err := is.FindClosestServer(); err != nil {
+			return fmt.Errorf("unable to find closest server: %s", err)
 		}
 	}
 
-	is.Log.Debug("Starting Speed Test")
-	is.Log.Debug("Running Ping...")
-	err := s.PingTest()
+	err := is.server.PingTest()
 	if err != nil {
 		return fmt.Errorf("ping test failed: %v", err)
 	}
-	is.Log.Debug("Running Download...")
-	err = s.DownloadTest(is.MemorySavingMode)
+	err = is.server.DownloadTest(is.MemorySavingMode)
 	if err != nil {
-		is.Log.Debug("try `memory_saving_mode = true` if this fails consistently")
-		return fmt.Errorf("download test failed: %v", err)
+		return fmt.Errorf("download test failed, try `memory_saving_mode = true` if this fails consistently: %v", err)
 	}
-	is.Log.Debug("Running Upload...")
-	err = s.UploadTest(is.MemorySavingMode)
+	err = is.server.UploadTest(is.MemorySavingMode)
 	if err != nil {
-		is.Log.Debug("try `memory_saving_mode = true` if this fails consistently")
-		return fmt.Errorf("upload test failed failed: %v", err)
+		return fmt.Errorf("upload test failed failed, try `memory_saving_mode = true` if this fails consistently: %v", err)
 	}
 
-	is.Log.Debug("Test finished.")
-
-	fields := make(map[string]interface{})
-	fields["download"] = s.DLSpeed
-	fields["upload"] = s.ULSpeed
-	fields["latency"] = timeDurationMillisecondToFloat64(s.Latency)
-
-	tags := make(map[string]string)
+	fields := map[string]any{
+		"download": is.server.DLSpeed,
+		"upload":   is.server.ULSpeed,
+		"latency":  timeDurationMillisecondToFloat64(is.server.Latency),
+	}
+	tags := map[string]string{
+		"server_id": is.server.ID,
+		"host":      is.server.Host,
+	}
 
 	acc.AddFields(measurement, fields, tags)
 	return nil
 }
 
+func (is *InternetSpeed) FindClosestServer() error {
+	user, err := speedtest.FetchUserInfo()
+	if err != nil {
+		return fmt.Errorf("fetching user info failed: %v", err)
```

Comment:
```suggestion
		return fmt.Errorf("fetching user info failed: %w", err)
```
Same for other instances. It's really better to wrap the error just in case we ever want to reason about it in the caller code...